libr/bin/p/bin_dis.c

```diff
@@ -0,0 +1,302 @@
+	}
+
+	if (!dis_read_operand (buf, &o->header.runtime_flags)) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.stack_extent)) {
+		goto invalid;
+	}
+	if (o->header.stack_extent < 0) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.code_size)) {
+		goto invalid;
+	}
+	if (o->header.code_size < 0) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.data_size)) {
+		goto invalid;
+	}
+	if (o->header.data_size < 0) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.type_size)) {
+		goto invalid;
+	}
+	if (o->header.type_size < 0) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.link_size)) {
+		goto invalid;
+	}
+	if (o->header.link_size < 0) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.entry_pc)) {
+		goto invalid;
+	}
+	if (o->header.entry_pc < 0) {
+		goto invalid;
+	}
+
+	if (!dis_read_operand (buf, &o->header.entry_type)) {
+		goto invalid;
+	}
+	if (o->header.entry_type < 0) {
+		goto invalid;
+	}
+
+	o->header_size = r_buf_tell (buf);
+
+	// in order to determine the size of the code section, we must parse
+	// all instructions. if we encounter an invalid instruction we must fail
+	// because instruction sizes are variable.
+	ut64 addr = r_buf_tell (buf);
+	for (i = 0; i < o->header.code_size; i++) {
+		struct dis_instr instr = {0};
+		ht_uu_insert (o->pcs, i, r_buf_tell (buf));
+		if (!dis_read_instr (buf, &instr)) {
+			R_LOG_ERROR ("Bad Dis instruction (pc is 0x%x)", i);
+			goto invalid;
+		}
+	}
+	o->code_size = r_buf_tell (buf) - addr;
+
+	// parse type section
+	addr = r_buf_tell (buf);
+	for (i = 0; i < o->header.type_size; i++) {
+		struct dis_type typ = {0};
+		if (!dis_read_type (buf, &typ)) {
+			goto invalid;
+		}
+	}
+	o->type_size = r_buf_tell (buf) - addr;
+
+	addr = r_buf_tell (buf);
+	// module name (ignored)
+	for (i = 0; ; i++) {
+		ut8 b;
+		if (r_buf_read (buf, &b, sizeof (b)) != sizeof (b)) {
+			goto invalid;
+		}
+		if (b == 0) {
+			break;
+		}
+	}
+	o->module_name_size = r_buf_tell (buf) - addr;
+
+	// parse link section
+	addr = r_buf_tell (buf);
+	for (i = 0; i < o->header.link_size; i++) {
+		struct dis_link link = {0};
+		if (!dis_read_link (buf, &link)) {
+			goto invalid;
+		}
+	}
+	o->link_size = r_buf_tell (buf) - addr;
+
+	if (bin_obj) {
+		*bin_obj = o;
+	}
+
+	return true;
+invalid:
+	ht_uu_free (o->pcs);
+	free (o);
+	return false;
+}
+
+static void destroy(RBinFile *bf) {
+	RBinDisObj *o = (RBinDisObj *)bf->o->bin_obj;
+	ht_uu_free (o->pcs);
+	free (o);
+}
+
+static RList *entries(RBinFile *bf) {
+	bool found;
+	RList *ret;
+	RBinAddr *ptr = NULL;
+	RBinDisObj *o = (RBinDisObj *)bf->o->bin_obj;
+
+	ut64 entry_address = ht_uu_find (o->pcs, o->header.entry_pc, &found);
+	if (!found) {
+		return NULL;
+	}
+
+	if (!(ret = r_list_new ())) {
+		return NULL;
+	}
+
+	ret->free = free;
+
+	if ((ptr = R_NEW0 (RBinAddr))) {
+		ptr->paddr = entry_address;
+		ptr->vaddr = entry_address;
```

Comment:
maybe vaddr should honor the baddr, but as long as its a vm i guess that its fine tha pa == va

Reply:
I don't implement `baddr` so it is 0... am I supposed to read the user-supplied base address somewhere for this?